src/functions.rs

```diff
@@ -1508,6 +1527,7 @@ pub struct FunctionInfo<'a> {
     // This will be initialized to true and prevalidate_functions will set this
     // to false if needed.
     pub is_castable: bool,
+    pub callers: BTreeSet<CallerInfo>,
```

Comment:
Would this have a chain of all callers (ancestors) of a given function or just the immediate callers(parents)?

I ask because if it is immediate callers, we may be able to use the caller name to map it to another function info, which may make the recursive check more efficient.

Reply:
From the propagate code below it looks like its just the parents.

If that is the case I think it should be "easy" to get a call stack and just make sure the function we are validating doesn't appear in it.

Reply:
Awesome, that would be a good optimization. We'll need to make sure that nothing in validate_functions() relies on the assumption that there's no recursion, since we'd be moving the check from prevalidate to postvalidate.